WMF Framework/Event Platform Client/EventPlatformClientLibrary.swift

```diff
@@ -0,0 +1,605 @@
+         *
+         * Stream cc-ing should be done only 1 level deep to avoid duplication.
+         * For example, edits shouldn't cc edits.growth AND edits.growth.test,
+         * since edits.growth.test would be cc'd by edits.growth.
+         *
+         * Instead of working stream by stream to find children streams for each
+         * parent stream, we work backwards by constructing parent streams from
+         * children streams. This enables children streams such as
+         * 'edits.growth' to be cc'd when logging events to 'edits' even when
+         * parent stream 'edits' is not present in the stream configuration.
+         *
+         * Refer to mw:Wikimedia_Product/Analytics_Infrastructure/Stream_configuration#Stream_cc-ing
+         * for additional documentation.
+         */
+        for stream in config.keys {
+            let s = stream.split(separator: ".")
+            let n_prefixes = s.count - 1
+            if n_prefixes > 1 {
+                for i in 1...n_prefixes {
+                    let child = s[0...i].joined(separator: ".")
+                    let parent = s[0..<i].joined(separator: ".")
+                    COPIED.append_if_new(key: parent, value: child)
+                }
+            } else if n_prefixes == 1 {
+                COPIED.append_if_new(key: String(s[0]), value: stream)
+            }
+        }
+        print("[EPC] Stream cc-ing:\n\(COPIED.prettyPrintJSON)")
+        if CONFIG != nil && input_buffer.count > 0 {
+            var cached_event: (stream: String, schema: String, data: [String: Any], domain: String?)? = input_buffer.remove(at: 0)
+            while cached_event != nil {
+                log(stream: cached_event!.stream,
+                    schema: cached_event!.schema,
+                    data: cached_event!.data, domain: cached_event!.domain)
+                cached_event = input_buffer.remove(at: 0) // next
+            }
+        }
+    }
+
+    /**
+     * Yields a deterministic (not stochastic) determination of whether the provided `id` is
+     * in-sample or out-of-sample according to the `acceptance` rate
+     * - Parameter id: either session ID generated with `generate_id` or the app install ID
+     * generated with `UUID().uuidString`
+     * - Parameter acceptance: the desired proportion of many `token`-s being accepted
+     *
+     * The algorithm works in a "widen the net on frozen fish" fashion -- tokens continue evaluating to
+     * true as the acceptance rate increases. For example, a device determined to be in-sample for a
+     * stream "A" having rate 0.1 will be determined to be in-sample for a stream "B" having rate 0.2,
+     * and its events will show up in tables "A" and "B".
+     */
+    private func determine(_ id: String, _ acceptance: Double) -> Bool {
+        let token: UInt32 = UInt32(id.prefix(8), radix: 16)!
+        return (Double(token) / Double(UInt32.max)) < acceptance
```

Comment:
Small suggestion to avoid the `!`:

```suggestion
        guard let token = UInt32(id.prefix(8), radix: 16) else {
            return false
        }
        
        return (Double(token) / Double(UInt32.max)) < acceptance
```